README.md
# Zway-DeviceMove

Move devices like blinds or windows based on timing information to a selected
position.

Some devices cannot be set to a specific position, but only can be controlled
by moving them up or down. This module tries to circumvent this limitation by
creating a virtual multilevel switch device instead of the original device,
that translates the user-select position into timing information.

For example if a device usually takes 40 seconds to move from fully closed to
fully opened state, then moving the device to 50% requires sending the
move-up command, followed by a stop command 20 seconds later. The module also
tries to detect manual movement of the device.

Furthermore this virtual device can check related devices. When two automation
devices operate on the same physical object (eg. blind and window chain drive)
, and interfere witch each other then checking for related devices ensures
proper functionality and prevents damage.

# Configuration

## devices.device

Managed devices

## devices.timeUp

Time it takes to move from fully closed to fully opened

## devices.timeDown

Time it takes to move from fully opened to fully closed

## devices.relatedDevice

Related device. See relatedCheck for detailed explanation.

## difference

Do not move device when difference between current and target position is less than this value

## icon

Select which icon to use for the virtual device. Default (taken from the
parent device), blinds, windows and dimmer icons are available.

## report

Indicates at which position correct device positioning is reported.
Eg. A blind only reports 0% when fully closed, or 100% when partially or fully
opened. In this case the report setting needs to be set to report-'close',
because only the fully closed state is reported correctly.

## relatedCheck

Enables checks for related devices. When two automation devices operate
on the same physical object, and interfere witch each other then checking
for related devices ensures proper functionality and prevents damage.
Eg. Closing a blind is not possible when the window was opened by a chain
drive. On the other hand windows should not be opened above a certain
limit with closed blinds. This check is able to limit the movement of the
managed devices accordingly.

## relatedDeviceLimit, relatedDeviceComparison

Checks if the related device is above or below a certain limit. If
relatedDeviceLimit and relatedDeviceComparison match then the current device
is limited by the deviceLimit

## deviceLimit

Defines a limit for the current device if the related device position is
within a certain limit.

# Events

No events are emitted

# Virtual Devices

This module creates a virtual device for each managed device. The original
device is hidden.

# Installation

Install the BaseModule from https://github.com/maros/Zway-BaseModule first

The prefered way of installing this module is via the "Zwave.me App Store"
available in 2.2.0 and higher. For stable module releases no access token is
required. If you want to test the latest pre-releases use 'k1_beta' as
app store access token.

For developers and users of older Zway versions installation via git is
recommended.

```shell
cd /opt/z-way-server/automation/userModules
git clone https://github.com/maros/Zway-DeviceMove.git DeviceMove --branch latest
```

To update or install a specific version
```shell
cd /opt/z-way-server/automation/userModules/DeviceMove
git fetch --tags
# For latest released version
git checkout tags/latest
# For a specific version
git checkout tags/1.02
# For development version
git checkout -b master --track origin/master
```

# License

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or any
later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.


CHANGES
1.10 2019 Apr 26, Fri
    * Fix handling of target key
    * Reset metrics:target once level is at target
    * Improve handling of related devices
    * Unify handling of min/max level
    * Merge pull request #2 from joel-ryden/master
    * Set correct level and clear locks after stop command
    * Send stop command to device even when in max states
    * Set new level after movement is finished
    * Update module documentation
    * Update copyright
    * Prevent race condition when device update would check a device during movement or while being locked after user interactio

1.09 2016 Oct 22, Sat
    * Calculate approximate position on device stop
    * Shorter lock times

1.08 2016 Jul 03, Sun
    * Fix difference check
    * Configurable steps
    * Delayed fix movement
    * Set min/max

1.07 2016 Jun 07, Tue
    * Fix target level handling
    * Improve/add labels/help texts
    * Add configurable difference
    * Set defaults
    * Delay related check on init since other device might not be ready

1.06 2016 Apr 20, Wed
    * Comparison strict and lax mode
    * Do not move device if already beyond related device limit
    * Remove special handling of window icons
    * Set probeType
    * Ignore levels > 100 when checking for target level
    * Fix log message
    * Remove useless and broken code
    * Store device target and move there once possible
    * Update icon name
    * Set target level
    * Set back location and tags to real device
    * Set maxdiff globally
    * Fix check for definedness
    * Add basic stop command
    * Update installation howto
    * Update readme

1.05 2016 Jan 22, Fri
    * Add doc link
    * Rename title/description key to make module work in appstore

1.04 2016 Jan 02, Sat
    * Update documentation
    * Fix related checks
    * Improve logs

1.03 2015 Dec 27, Sun
    * Use BaseModule
    * Convert to int one step later
    * Add different commands for full up/down
    * Add more logging
    * No delay for full up/down

1.02 2015 Dec 20, Sun
    * Update title
    * German translation
    * Handle different up and down move times
    * Add missing command
    * No delay for full up/down
    * Ensure max level
    * Read from probeType instead of metrics: probeType
    * Level 255 = fully opened
    * Poll device on update
    * Allow to add binary sensors as related devices
    * Fix device creation
    * Update documentation
    * Various fixups

1.01 2015 Oct 25, Sun
    * Fix visibility
    * Add documentation

1.00 2015 Oct 17, Sat
    * Initial Release 


timeout.js
/*jshint -W058 */

function TimeoutManager(scope) {
    var self        = this;
    self.scope      = scope;
    self.timeouts   = {};
}
TimeoutManager.prototype.timeouts = {};
TimeoutManager.prototype.running  = function(id) {
    var self   = this;
    if (typeof(self.timeouts[id]) === "undefined") {
        return false;
    }
    return self.timeouts[id].cleared ? false:true;
};
TimeoutManager.prototype.add      = function(id,fn,interval) {
    var self   = this;

    // Do not overwrite existing timer
    if (typeof(self.timeouts[id]) !== "undefined"
        && self.timeouts[id].cleared === false) {
        throw('Timeout "'+id+'" is already active');
    }

    // Build args
    var args   = new Array(arguments.length - 1);
    for(var i = 0; i < args.length; ++i) {
        args[i] = arguments[i+1];
    }
    args.unshift(self.scope,self.scope);
    self.timeouts[id] = new(Function.prototype.bind.apply(Timeout,args));
    return self.timeouts[id];
};
TimeoutManager.prototype.replace  = function(id,fn,interval) {
    var self   = this;

    // Clear existing timer
    if (typeof(self.timeouts[id]) !== "undefined") {
        self.timeouts[id].clear();
    }

    // Build args
    var args   = new Array(arguments.length - 1);
    for(var i = 0; i < args.length; ++i) {
        args[i] = arguments[i+1];
    }
    args.unshift(self.scope,self.scope);
    self.timeouts[id] = new(Function.prototype.bind.apply(Timeout,args));
    return self.timeouts[id];
};
TimeoutManager.prototype.get    = function(id) {
    return this.timeouts[id];
};
TimeoutManager.prototype.clear    = function(id) {
    var self   = this;
    if (typeof(self.timeouts[id]) !== "undefined") {
        self.timeouts[id].clear();
        delete self.timeouts[id];
        return true;
    }
    return false;
};
TimeoutManager.prototype.clearAll = function() {
    var self   = this;
    for (var id in self.timeouts) {
        if (self.timeouts.hasOwnProperty(id)  
            && typeof(self.timeouts[id]) !== "undefined") {
            self.timeouts[id].clear();
        }
    }
    self.timeouts = {};
};


function Timeout(scope,fn,interval) {
    var self   = this;
    var args   = new Array(arguments.length-3);
    for(var i = 0; i < args.length; ++i) {
        args[i] = arguments[i+3];
    }
    args.unshift(scope);

    self.fn     = Function.prototype.bind.apply(fn,args);
    self.id     = setTimeout(self.run.bind(self),interval);
    self.cleared= false;
    //Register timeout by name?
}

Timeout.prototype.id       = undefined;
Timeout.prototype.cleared  = false;
Timeout.prototype.fn       = undefined;
Timeout.prototype.run      = function() {
    this.clear();
    this.fn();
};
Timeout.prototype.clear    = function() {
    var self   = this;
    if (typeof(self.id) !== 'undefined' && self.cleared === false) {
        clearTimeout(self.id);
    }
    self.cleared = true;
    self.id = undefined;
};




smarttimeout.js
function TimeoutManager(scope) {
    var self   = this;
    self.scope = scope;
}
TimeoutManager.prototype.timeouts = {};
TimeoutManager.prototype.add      = function(id,fn,interval) {
    var self   = this;

    // Do not overwrite existing timer
    if (typeof(self.timeouts[id]) !== "undefined"
        && self.timeouts[id].cleared === false) {
        throw('Timeout '+id+' is already active');
    }

    // Build args
    var args   = new Array(arguments.length - 1);
    for(var i = 0; i < args.length; ++i) {
        args[i] = arguments[i+1];
    }
    args.unshift(self.scope,self.scope);
    
    return self.timeouts[id] = new(Function.prototype.bind.apply(Timeout,args));
};
TimeoutManager.prototype.replace  = function(id,fn,interval) {
    var self   = this;

    // Clear existing timer
    if (typeof(self.timeouts[id]) !== "undefined") {
        self.timeouts[id].clear();
    }

    // Build args
    var args   = new Array(arguments.length - 1);
    for(var i = 0; i < args.length; ++i) {
        args[i] = arguments[i+1];
    }
    args.unshift(self.scope);

    return self.timeouts[id] = new(Function.prototype.bind.apply(Timeout,args));
};
TimeoutManager.prototype.clear    = function(id) {
    var self   = this;
    if (typeof(self.timeouts[id]) !== "undefined") {
        self.timeouts[id].clear();
        delete self.timeouts[id];
        return true;      
    }
    return false;
};
TimeoutManager.prototype.clearAll = function() {
    var self   = this;
    for (var id in self.timeouts) {
        if (timeout.hasOwnProperty(id) 
            && typeof(self.timeouts[id]) !== "undefined") {
            self.timeouts[id].clear();
        }
    }
    self.timeouts = {};
};


function Timeout(scope,fn,interval) {
    var self   = this;
    var args   = new Array(arguments.length-3);
    for(var i = 0; i < args.length; ++i) {
        args[i] = arguments[i+3];
    }
    args.unshift(scope);

    self.fn    = Function.prototype.bind.apply(fn,args);
    self.id    = setTimeout(self.run.bind(self),interval);
    //Register timeout by name?
}

Timeout.prototype.id       = null;
Timeout.prototype.cleared  = false;
Timeout.prototype.fn       = null;
Timeout.prototype.run      = function() {
    this.clear();
    this.fn();
};
Timeout.prototype.clear    = function() {
    if (typeof(this.id) !== 'null' && this.cleared == false) {
        clearTimeout(this.id);
    }
    this.cleared = true;
    this.id = null;
};




index.js
/*** DeviceMove Z-Way HA module *******************************************

Version: 1.10
(c) Maroš Kollár, 2015-2017
-----------------------------------------------------------------------------
Author: Maroš Kollár <maros@k-1.com>
Description:
    Move devices to specified position based on timing information

******************************************************************************/

function DeviceMove (id, controller) {
    // Call superconstructor first (AutomationModule)
    DeviceMove.super_.call(this, id, controller);

    this.virtualDevices = {};
    this.callbacks      = {};
    this.delay          = undefined;
    this.lock           = undefined;
    this.timer          = undefined;
}

inherits(DeviceMove, BaseModule);

_module = DeviceMove;

// ----------------------------------------------------------------------------
// --- Module instance initialized
// ----------------------------------------------------------------------------

DeviceMove.prototype.init = function (config) {
    DeviceMove.super_.prototype.init.call(this, config);
    var self = this;

    self.difference     = parseInt(self.config.difference || 10);
    self.step           = parseInt(self.config.step || 1);
    self.delay          = new TimeoutManager(self);
    self.lock           = new TimeoutManager(self);
    self.timer          = setInterval(
        _.bind(self.checkAllDevices,self),
        (5*60*1000)
    );

    setTimeout(_.bind(self.initCallback,self),10000);
};

DeviceMove.prototype.initCallback = function() {
    var self = this;

    var icon = self.config.icon;

    // Loop all devices
    _.each(self.config.devices,function(deviceEntry) {
        var deviceId    = deviceEntry.device;
        var realDevice  = self.controller.devices.get(deviceId);
        if (realDevice === null) {
            self.error('Device not found '+deviceId);
            return;
        }
        // Set icon
        var deviceIcon  = icon;
        //var probeType   = realDevice.get('probeType');
        var title       = realDevice.get('metrics:title');
        if (icon === 'default') {
            deviceIcon  = realDevice.get('metrics:icon');
        } else if (icon === 'blind') {
            deviceIcon = 'blinds';
        } else {
            deviceIcon = icon;
        }

        // Hide and rename device
        title = title.replace(/\s*\[raw\]/,"");
        realDevice.set('metrics:title',title+' [raw]');
        realDevice.set('visibility', false);

        // Create virtual device
        var virtualDevice = this.controller.devices.create({
            deviceId: "DeviceMove_" + self.id+'_'+deviceId,
            defaults: {
                metrics: {
                    title: title,
                    level: null
                }
            },
            overlay: {
                deviceType: 'switchMultilevel',
                probeType: 'motor',
                tags: realDevice.get('tags'),
                location: realDevice.get('location'),
                metrics: {
                    icon: deviceIcon,
                    min: 0,
                    max: 99,
                    step: parseInt(self.config.step || 1)
                }
            },
            handler: function(command,args) {
                if (command === 'update') {
                    self.pollDevice(deviceId);
                    return;
                }
                var currentLevel = Math.min(99,this.get('metrics:level'));
                var step = parseInt(self.config.step || 1);
                var newLevel;
                var delay = false;
                if (command === 'on' || command === 'up' || command === 'startUp' || command === 'upMax') {
                    newLevel = 99;
                } else if (command === 'off'|| command === 'down' || command === 'startDown') {
                    newLevel = 0;
                } else if ("exact" === command || "exactSmooth" === command) {
                    newLevel = args.level;
                    newLevel = Math.round(newLevel / step) * step;
                    if (newLevel === 0 && args.level > 0) {
                        newLevel = Math.max(step,10);
                    }
                    delay = true;
                } else if ("increase" === command) {
                    newLevel = currentLevel + step;
                } else if ("decrease" === command) {
                    newLevel = currentLevel - step;
                } else if ("stop" === command) {
                    // TODO figure out if we are currently moving, and try to calc new position
                    var action = this.get('metrics:action');
                    self.log('Stopped device '+deviceId);
                    if (_.isObject(action)) {
                        var currentTime = new Date().getTime() / 1000;
                        var oldLevel    = Math.min(99,action.oldLevel);
                        var diffTime    = currentTime - action.startTime;
                        var stepTime    = deviceEntry[(oldLevel > currentLevel) ? 'timeDown':'timeUp'] / 100;
                        var diffLevel   = parseInt(diffTime / stepTime,10);
                        var stopLevel   = (oldLevel > currentLevel) ? (oldLevel - diffLevel) : (oldLevel + diffLevel);
                        stopLevel = Math.min(99, stopLevel);
                        stopLevel = Math.max(0, stopLevel);
                        self.log('Moved '+diffTime+'sec ('+stepTime+'sec/step) to '+stopLevel+' (from '+oldLevel+')');
                        this.set('metrics:action',null,{ silent: true });
                        this.set('metrics:target',null,{ silent: true });
                        this.set('metrics:level',stopLevel,{ silent: true });
                        self.delay.clear(deviceId);
                        self.lock.clear(deviceId);
                    }
                    var realDevice = self.controller.devices.get(deviceId);
                    realDevice.performCommand('stop');
                    return;
                }

                if (currentLevel === newLevel) {
                    //self.log('Not moving device '+deviceId+'. Same level');
                    return;
                }
                if (newLevel === 0 || newLevel >= 99) {
                    delay = false;
                }
                this.set('metrics:target',newLevel,{ silent: true });
                self.log('Got command '+command+' for '+deviceId+': Set from '+currentLevel+' to '+newLevel+((delay) ? ' with delay':' without delay'));
                if (delay) {
                    self.delay.replace(
                        deviceId,
                        self.moveDevice,
                        1000*1.5,
                        deviceId,
                        newLevel,
                        'manual delayed'
                    );
                } else {
                    self.moveDevice(deviceId,newLevel,'manual');
                }
            },
            moduleId: self.id
        });

        self.virtualDevices[deviceId] = virtualDevice;

        // Build, register and call check callback
        var callback = _.bind(self.checkDevice,self,deviceId);
        realDevice.on('change:metrics:level',callback);
        setTimeout(callback,1000*5);
        self.callbacks[deviceId] = callback;
    });
};

DeviceMove.prototype.stop = function() {
    var self = this;

    DeviceMove.super_.prototype.stop.call(this);

    // Remove device & callbacks
    _.each(self.config.devices,function(deviceId){
        var realDevice = self.controller.devices.get(deviceId);
        if (realDevice === null) {
            return;
        }
        var virtualDevice = self.controller.devices.get(self.virtualDevices[deviceId]);

        if (virtualDevice !== null) {
            // Write back changes to real device
            var title       = realDevice.get('metrics:title');
            title = title.replace(/\s*\[raw\]\s*/,"");
            realDevice.set('metrics:title',title);
            realDevice.set('visibility',true);
            realDevice.set('tags',virtualDevice.get('tags'));
            realDevice.set('location',virtualDevice.get('location'));

            self.controller.devices.remove(virtualDevice);
        }

        if (virtualDevice !== null) {
            realDevice.off('change:metrics:level',self.callbacks[deviceId]);
        }
    });

    if (typeof(self.timer) !== 'undefined') {
        clearInterval(self.timer);
    }

    self.delay.clearAll();
    self.lock.clearAll();

    self.timer          = undefined;
    self.delay          = undefined;
    self.lock           = undefined;
    self.virtualDevices = {};
    self.callbacks      = {};
};

// ----------------------------------------------------------------------------
// --- Module methods
// ----------------------------------------------------------------------------

DeviceMove.prototype.setStatus = function(deviceId,level) {
    var self            = this;
    var virtualDevice   = self.virtualDevices[deviceId];
    level               = parseInt(level,10);

    if (level > 99) {
        level = 99;
    }

    // Set virtual device
    virtualDevice.set('metrics:level',level);
};

DeviceMove.prototype.moveDevice = function(deviceId,level,source) {
    var self = this;

    self.log('Got move device '+ deviceId+' to '+level+' from '+source);

    // Check if already running
    if (self.lock.running(deviceId)) {
        self.log('Device '+ deviceId+' is locked. Delay action');
        self.delay.replace(
            deviceId,
            self.moveDevice,
            (1000*10),
            deviceId,
            level, // Can produce race condition!!
            source + ' delayed'
        );
        return;
    }

    var virtualDevice   = self.virtualDevices[deviceId];
    var oldLevel        = virtualDevice.get('metrics:level');
    oldLevel            = Math.min(oldLevel,99);
    var realDevice      = self.controller.devices.get(deviceId);
    var deviceEntry     = _.find(self.config.devices,function(deviceEntry) { return deviceEntry.device === deviceId; });
    if (deviceEntry === null) {
        self.error('Could not find real device '+deviceId);
        return;
    }
    var moveCommand;
    var newLevel        = parseInt(level,10);
    var maxTime         = Math.max(deviceEntry.timeUp,deviceEntry.timeDown);
    var commandLevel    = newLevel;
    var currentTime     = new Date().getTime() / 1000;

    // Check related devices
    if (self.config.relatedCheck
        && typeof(deviceEntry.relatedDevice) !== 'undefined') {
        var relatedDevice = self.controller.devices.get(deviceEntry.relatedDevice);
        if (relatedDevice === null) {
            self.error('Related device for '+deviceId+' not found '+deviceEntry.relatedDevice);
        } else {
            // Get related device level
            var relatedLevel = relatedDevice.get('metrics:level');
            var relatedRestrict = false;
            if (typeof(relatedLevel) === 'string'
                && relatedLevel === 'on') {
                relatedLevel = 99;
            } else if (typeof(relatedLevel) === 'string'
                && relatedLevel === 'off') {
                relatedLevel = 0;
            } else {
                relatedLevel = parseInt(relatedLevel,10);
            }
            // Fix level
            relatedLevel = Math.min(relatedLevel,99);

            // Related level - gt
            if ((self.config.relatedDeviceComparison === 'gt' || self.config.relatedDeviceComparison === 'gt_strict')
                && relatedLevel > self.config.relatedDeviceLimit) {
                if (oldLevel < self.config.deviceLimit
                    && self.config.relatedDeviceComparison === 'gt_strict') {
                    newLevel = oldLevel;
                } else {
                    newLevel = Math.max(newLevel,self.config.deviceLimit);
                }
                relatedRestrict = true;
            // Related level - lt
            } else if ((self.config.relatedDeviceComparison === 'lt' || self.config.relatedDeviceComparison === 'lt_strict')
                && relatedLevel < self.config.relatedDeviceLimit) {
                if (oldLevel > self.config.deviceLimit
                    && self.config.relatedDeviceComparison === 'lt_strict') {
                    newLevel = oldLevel;
                } else {
                    newLevel = Math.min(newLevel,self.config.deviceLimit);
                }
                relatedRestrict = true;
            }
            // Check new level
            if (relatedRestrict
                && newLevel === oldLevel) {
                self.log('Not movining due to related device at '+relatedLevel);
                return;
            } else if (commandLevel !== newLevel) {
                self.log('Constrained level to '+newLevel+' due to related device at '+relatedLevel);
            }
        }
    }

    if (newLevel >= 99) {
        moveCommand = 'startUp';
        newLevel = 255;
        self.lock.add(
            deviceId,
            function() {
                self.pollDevice(deviceId);
                virtualDevice.set('metrics:level',newLevel);
                self.stopDevice(deviceId);
            },
            (maxTime*1.1*1000)
        );
        virtualDevice.set('metrics:action',{
            startTime:  currentTime,
            oldLevel:   oldLevel,
            targetLevel:99,
            runTime:    deviceEntry.timeUp
        });
        realDevice.set('metrics:level',255);
    } else if (newLevel <= 0) {
        moveCommand = 'startDown';
        newLevel = 0;
        self.lock.add(
            deviceId,
            function() {
                self.pollDevice(deviceId);
                virtualDevice.set('metrics:level',newLevel);
                self.stopDevice(deviceId);
            },
            (maxTime*1.1*1000)
        );
        virtualDevice.set('metrics:action',{
            startTime:  currentTime,
            oldLevel:   oldLevel,
            targetLevel:0,
            runTime:    deviceEntry.timeDown
        });
        realDevice.set('metrics:level',0);
    } else {
        var diffLevel = Math.abs(oldLevel - newLevel);
        if (diffLevel < self.difference && oldLevel !== 0) {
            self.log('Not movining due minimum difference');
            return;
        }
        var stepTime    = deviceEntry[(oldLevel > newLevel) ? 'timeDown':'timeUp'] / 100;
        var diffTime    = stepTime * diffLevel;
        diffLevel       = Math.abs(diffTime / stepTime);
        moveCommand     = (oldLevel > newLevel) ? 'startDown':'startUp';
        newLevel        = parseInt((oldLevel < newLevel) ? oldLevel + diffLevel : oldLevel - diffLevel,10);
        self.lock.add(
            deviceId,
            function() {
                self.stopDevice(deviceId);
                virtualDevice.set('metrics:level',newLevel);
            },
            (diffTime * 1000)
        );
        virtualDevice.set('metrics:action',{
            startTime:  currentTime,
            oldLevel:   oldLevel,
            targetLevel:newLevel,
            runTime:    diffTime
        });
        self.log('Move device '+deviceId+' from '+oldLevel+' to '+newLevel+' for '+diffTime+' seconds');
    }

    realDevice.performCommand(moveCommand);
    self.setStatus(deviceId,newLevel);
};

DeviceMove.prototype.stopDevice = function(deviceId) {
    var self        = this;
    var realDevice  = self.controller.devices.get(deviceId);
    self.lock.add(
        deviceId,
        self.checkDevice,
        (5*1000),
        deviceId
    );

    self.virtualDevices[deviceId].set('metrics:action',null);
    self.virtualDevices[deviceId].set('metrics:target',null,{ 'silent': true });

    realDevice.performCommand("stop");
    realDevice.performCommand("update");
    //self.pollDevice(deviceId);
};

DeviceMove.prototype.checkAllDevices = function() {
    var self = this;
    self.log('Checking devices');

    _.each(self.config.devices,function(deviceEntry) {
        var deviceId        = deviceEntry.device;
        var realDevice      = self.controller.devices.get(deviceId);
        var virtualDevice   = self.virtualDevices[deviceId];
        var virtualLevel    = parseInt(virtualDevice.get('metrics:level'),10);
        var targetLevel     = virtualDevice.get('metrics:target');
        if (self.lock.running(deviceId)) {
            return;
        }
        if (typeof(targetLevel) === undefined || targetLevel === null || isNaN(targetLevel)) {
            return;
        } else {
            targetLevel = parseInt(targetLevel,10);
            if (targetLevel === virtualLevel) {
                virtualDevice.set('metrics:target',null,{ silent: true });
                return;
            }
        }

        virtualLevel        = Math.min(virtualLevel,100);
        targetLevel         = Math.min(targetLevel,100);
        self.log('Checking device'+deviceId+': '+virtualLevel+'-'+targetLevel+':'+Math.abs(virtualLevel - targetLevel));

        // Set target level
        if (virtualLevel !== targetLevel
            && Math.abs(virtualLevel - targetLevel) >= self.difference) {
            setTimeout(function() {
                self.log('Detected target mismatch for '+deviceId+'. Now moving');
                self.moveDevice(deviceId,targetLevel,'mismatch');
            },1000*30);
        }
        self.pollDevice(deviceId);
    });
};

DeviceMove.prototype.pollDevice = function(deviceId) {
    var self = this;

    var pollInterval    = 10*60*1000;
    var currentTime     = Math.floor(new Date().getTime() / 1000);
    var realDevice      = self.controller.devices.get(deviceId);
    if (realDevice === null) {
        return;
    }
    var updateTime      = realDevice.get('updateTime');
    if ((updateTime + pollInterval) < currentTime) {
        if (self.lock.running(realDevice.id)) {
            return;
        }
        self.log('Polling device '+realDevice.id);
        realDevice.performCommand("update");
    }
};

DeviceMove.prototype.checkDevice = function(deviceId,args) {
    var self            = this;

    if (self.lock.running(deviceId)) {
        return;
    }

    var realDevice      = self.controller.devices.get(deviceId);
    var virtualDevice   = self.virtualDevices[deviceId];
    var realLevel       = parseInt(realDevice.get('metrics:level'),10);
    var virtualLevel    = parseInt(virtualDevice.get('metrics:level'),10);
    var targetLevel     = virtualDevice.get('metrics:target');
    virtualLevel        = Math.min(virtualLevel,100);
    var setLevel;

    // Detect full open
    if (self.config.report === 'open' && realLevel >= 99) {
        self.setStatus(deviceId,99);
    // Detect full close
    } else if (self.config.report === 'close' && realLevel === 0) {
        self.setStatus(deviceId,0);
    // Init empty slot
    } else if (isNaN(virtualLevel)) {
        self.setStatus(deviceId,realLevel);
    // Correct partial open
    } else if (self.config.report === 'close' && realLevel > 0 && virtualLevel === 0) {
        self.log('Detected status mismatch for '+deviceId+'. Now closed. Reporting '+realLevel);
        self.setStatus(deviceId,realLevel);
    // Correct partial close
    } else if (self.config.report === 'open' && realLevel === 0 && virtualLevel >= 99) {
        self.log('Detected status mismatch for '+deviceId+'. Now opened. Reporting '+realLevel);
        self.setStatus(deviceId,realLevel);
    }

    // Set update time
    virtualDevice.set('updateTime',realDevice.get('updateTime'));

    // Check target level
    if (targetLevel !== null) {
        targetLevel = Math.min(parseInt(targetLevel,10),100);
        if (virtualLevel !== targetLevel
            && Math.abs(virtualLevel - targetLevel) >= self.diff) {
            self.log('Detected target mismatch for '+deviceId+'. Now moving');
            self.moveDevice(deviceId,targetLevel,'check');
        } else if (virtualLevel === targetLevel) {
            virtualDevice.set('metrics:target',null,{ silent: true });
        }
    }
};


